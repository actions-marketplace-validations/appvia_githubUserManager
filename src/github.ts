import { createAppAuth } from '@octokit/auth-app'
import { Octokit } from '@octokit/rest'
const ignoredUsers = process.env.IGNORED_USERS.toLowerCase().split(',')

const octokit = getAuthenticatedOctokit()

// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
export function getAuthenticatedOctokit() {
  return new Octokit({
    authStrategy: createAppAuth,
    auth: {
      appId: process.env.GITHUB_APP_ID,
      privateKey: Buffer.from(process.env.GITHUB_PRIVATE_KEY, 'base64').toString('utf-8'),
      installationId: process.env.GITHUB_INSTALLATION_ID,
    },
  })
}

export async function getGithubUsersFromGithub(): Set<string> {
  const members = await octokit.paginate(octokit.orgs.listMembers, {
    org: process.env.GITHUB_ORG,
  })

  const pendingInvites = await octokit.paginate(octokit.orgs.listPendingInvitations, {
    org: process.env.GITHUB_ORG,
  })
  const pendingGithubAccounts = formatUserList(pendingInvites)

  const githubAccounts = formatUserList(members)

  if (pendingGithubAccounts.size > 0)
    console.log(`Outstanding GitHub invites for ${Array.from(pendingGithubAccounts).join(', ')}`)

  return new Set([...githubAccounts, ...pendingGithubAccounts])
}

// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
export function formatUserList(users): Set<string> {
  return new Set(users.map((user) => user.login.toLowerCase()))
}

export async function getUserIdFromUsername(username: string): Promise<number> {
  console.log(`Looking up user ${username}`)
  let user
  try {
    user = await octokit.users.getByUsername({ username })
  } catch (error) {
    throw `Unable to find user id for ${username}`
  }
  console.log(`User ${username} userid: ${user.data.id}`)
  return user.data.id
}

export async function addUsersToGitHubOrg(users: Set<string>): Promise<void> {
  for (const user of users) {
    await addUserToGitHubOrg(user)
  }
}

export async function addUserToGitHubOrg(user: string): Promise<void> {
  if (ignoredUsers.includes(user.toLowerCase())) {
    console.log(`Ignoring add for ${user}`)
    return false
  }
  const userId = await getUserIdFromUsername(user)
  console.log(`Inviting ${user} (${userId} to ${process.env.GITHUB_ORG})`)
  await octokit.orgs.createInvitation({
    org: process.env.GITHUB_ORG,
    invitee_id: userId,
  })
  console.log(`Invitation sent to ${user} (${userId} to ${process.env.GITHUB_ORG})`)
}

export async function removeUsersToGitHubOrg(users: Set<string>): Promise<void> {
  for (const user of users) {
    await removeUserToGitHubOrg(user)
  }
}

// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
export async function removeUserToGitHubOrg(user: string) {
  if (ignoredUsers.includes(user.toLowerCase())) {
    console.log(`Ignoring remove for ${user}`)
    return false
  }
  console.log(`Removing user/invitation ${user} from ${process.env.GITHUB_ORG}`)
  return octokit.orgs.removeMembershipForUser({
    org: process.env.GITHUB_ORG,
    username: user,
  })
}
