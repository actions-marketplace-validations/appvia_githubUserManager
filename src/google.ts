const { google } = require('googleapis')

const privatekey = JSON.parse(
  Buffer.from(process.env.GOOGLE_CREDENTIALS, 'base64').toString('utf-8')
)

export const googleAuth = async () => {
  const jwtClient = new google.auth.JWT(
    privatekey.client_email,
    null,
    privatekey.private_key,
    ['https://www.googleapis.com/auth/admin.directory.user.readonly'],
    process.env.GOOGLE_EMAIL_ADDRESS
  )
  await jwtClient.authorize()
  return jwtClient
}

export async function getAdminService () {
  return google.admin({
    version: 'directory_v1',
    auth: await googleAuth()
  })
}

export async function getGithubUsersFromGoogle () {
  const service = await getAdminService()

  const userList: {
    data: { users: Array<userResponseEntry> }
  } = await service.users.list({
    customer: 'my_customer',
    maxResults: 250,
    projection: 'custom',
    fields: 'users(customSchemas/Accounts/github(value))',
    customFieldMask: 'Accounts'
  })

  const githubAccounts = formatUserList(userList.data.users)
  return githubAccounts
}

interface userResponseEntry {
  customSchemas: { Accounts: { github: Array<{ value: string }> } }
}

export function formatUserList (users: Array<userResponseEntry>): Set<string> {
  return new Set(
    users
      .map(user =>
        user?.customSchemas?.Accounts?.github.map(account =>
          account.value.toLowerCase()
        )
      )
      .flat()
      .filter(Boolean)
  )
}
