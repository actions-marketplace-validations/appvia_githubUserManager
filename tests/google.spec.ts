jest.mock('googleapis')
import { google } from 'googleapis'
import * as mod from '../src/google'

const fakeUsersResponse = [
  { customSchemas: { Accounts: { github: [{ value: 'chrisns' }] } } },
  {
    customSchemas: {
      Accounts: { github: [{ value: 'Foo' }, , { value: 'tar' }] },
    },
  },
  {
    customSchemas: {
      Accounts: { github: [{ value: 'foo' }, { value: 'bar' }] },
    },
  },
]

describe('google integration', () => {
  beforeEach(() => {
    process.env.GOOGLE_EMAIL_ADDRESS = 'hello@example.com'
    process.env.GOOGLE_CREDENTIALS = Buffer.from(JSON.stringify({ client_email: 'foo', private_key: 'bar' })).toString(
      'base64',
    )
    jest.resetAllMocks()
  })
  it('googleAuth', () => {
    mod.googleAuth()
    // @ts-expect-error .mocks isn't on the original object so will fail
    return expect(google.auth.JWT.mock.calls).toMatchSnapshot()
  })
  it('getAdminService', () => {
    // @ts-expect-error mock service isn't a complete implementation, so being lazy and just doing the bare minimum
    google.admin.mockReturnValue('adminservice')
    const result = mod.getAdminService()
    return expect(result).resolves.toBe('adminservice')
  })

  it('getGithubUsersFromGoogle', () => {
    const service = { users: { list: jest.fn().mockResolvedValue({ data: { users: fakeUsersResponse } }) } }
    // @ts-expect-error mock service isn't a complete implementation, so being lazy and just doing the bare minimum
    jest.spyOn(mod, 'getAdminService').mockResolvedValue(service)

    const result = mod.getGithubUsersFromGoogle()
    expect(result).resolves.toMatchSnapshot()
  })

  it('formatUserList', () =>
    expect(mod.formatUserList(fakeUsersResponse)).toEqual(new Set(['chrisns', 'foo', 'bar', 'tar'])))
})
