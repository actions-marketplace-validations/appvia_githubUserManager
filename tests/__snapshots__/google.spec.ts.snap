// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`google integration formatUserList 1`] = `
Set {
  "chrisns",
  "foo",
  "tar",
  "bar",
}
`;

exports[`google integration formatUserList bad 1`] = `
Set {
  "chrisns",
}
`;

exports[`google integration getGithubUsersFromGoogle 1`] = `
Set {
  "chrisns",
  "foo",
  "tar",
  "bar",
}
`;

exports[`google integration googleAuth 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "foo",
      null,
      "bar",
      Array [
        "https://www.googleapis.com/auth/admin.directory.user.readonly",
      ],
      "hello@example.com",
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;
